### PR TITLE
Migrate build_resolvers

### PR DESCRIPTION
**Pull request by simolus3**
I missed the return type of `astNodeFor` in `package:build`, it should be nullable as indicated in the documentation.

### CODE REVIEW

File: build_resolvers/lib/src/resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/resolver.dart:342]: nit: We shouldn't need the `return` here.

File: build_resolvers/lib/src/resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/resolver.dart:469]:
```suggestion
FeatureSet _featureSet({List<String> enableExperiments = const []}) {
```

We used to use the old pattern as a defensive thing to guard against passing null, but now we can actually be sure we don't get it (well, assuming sound null safety but good enough).

File: build_resolvers/pubspec.yaml
Comment (jakemac53) [inline build_resolvers/pubspec.yaml:2]: We will want to bump to `2.0.0`

File: build_resolvers/test/resolver_test.dart
Comment (jakemac53) [inline build_resolvers/test/resolver_test.dart:5]: This is blocked on build_test right? Plan is to just come back and opt these in after landing both prs?

Comment (simolus3) [reply build_resolvers/test/resolver_test.dart:5]: I can migrate this test in the second PR, looks like I just missed it.

Comment (simolus3) [reply build_resolvers/test/resolver_test.dart:5]: Actually I already did: https://github.com/dart-lang/build/pull/3018/commits/273e80bdf687551880574fcccc5335cfdfbbbe5f

File: build_resolvers/lib/src/build_asset_uri_resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/build_asset_uri_resolver.dart:206]: We could do another `.whereType<String>()` to avoid the null check in the body here and the `!` in the following map function, but either way sgtm.

File: build_resolvers/lib/src/resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/resolver.dart:300]: It looks to me like this and `_packageConfig` could both be `late` and non-nullable as well?

Comment (simolus3) [reply build_resolvers/lib/src/resolver.dart:300]: In `reset()`, we use `_uriResolver?.reset()` so we need to know whether it was initialized or not.  `_packageConfig` could be set in the constructor or in `_ensureInitialized`, so we need to check that too (and we can't move the initializer into a `late` field declaration because its async). I don't think we can actually avoid the null checks here?

Comment (jakemac53) [reply build_resolvers/lib/src/resolver.dart:300]: Ya if we need to do null checks then we need to keep this as you have it, sgtm

File: build_resolvers/lib/src/build_asset_uri_resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/build_asset_uri_resolver.dart:71]: Can we replace the null check + cast with a `whereType<_AssetState>()`?

File: build_runner/test/server/build_updates_client/reloading_manager_test.dart
Comment (simolus3) [inline build_runner/test/server/build_updates_client/reloading_manager_test.dart:33]: In `graphs` 1.x, the `edges`  function in `stronglyConnectedComponents` is not allowed to return `null` anymore. That sounds reasonable to me, but it's tricky to catch in non-migrated packages.

I don't know if `build_runner/web/hot_reload_client.dart` needs changes too, is there an integration test to find out?

Comment (jakemac53) [reply build_runner/test/server/build_updates_client/reloading_manager_test.dart:33]: I don't think we have a hot reload test any longer in build_runner (I don't think we do any webdriver testing which this would require). Webdev is generally where tests like that live but it has its own hot module reloading script which is a bit different if I recall correctly.

I believe it wouldn't return null though, all nodes in the graph should have parents, or else they wouldn't be in the app at all. That assumes we do skip doing anything if you edit a module not depended on by the app.

Comment (jakemac53) [reply build_runner/test/server/build_updates_client/reloading_manager_test.dart:33]: Honestly we could just finally remove entirely the ability to hot reload from this package, we really expect people to go through webdev for that now and the implementation in this package is likely to bitrot.

**Comment by simolus3** [comment]
Too early for a ping, but from my end this looks ready now :)

**Review by jakemac53 — APPROVED** [review]

**Comment by jakemac53** [comment]
Thanks @simolus3 🥳

**Pull request merged**
